### PR TITLE
Fix race between failover and two-phase commit

### LINKED ISSUE #417: Two-phase commit fails when failover triggers and yields
Some roles may yield during `apply_config`. Others may take some time to finish applying configuration.

At that moment two-phase commit may race and return an error:

```log
15:29:18.067 [localhost:11002] cartridge.failover I> Failover triggered
15:29:18.067 [localhost:11002] cartridge.failover I> Instance state changed: RolesConfigured -> ConfiguringRoles
15:29:18.069 [localhost:11002] cartridge.failover I> Apply config to audit log storage
15:29:18.069 [localhost:11002] cartridge.failover I> Applying jobs config
15:29:18.073 [localhost:11001] http/127.0.0.1:51248 twophase.lua:193 W> Updating config clusterwide...
15:29:18.077 [localhost:11001] http/127.0.0.1:51248 twophase.lua:253 W> (2PC) Preparation stage...
15:29:18.085 [localhost:11001] http/127.0.0.1:51248 twophase.lua:272 E> Error preparing for config update at localhost:11002:
                                                                        Prepare2pcError: Instance state is ConfiguringRoles, can't apply config in this state
```

### PR DESCRIPTION
**Pull request by rosik**
Some roles may yield during `apply_config`. Others may take some time
to finish applying configuration.

At that moment two-phase commit may race and return an error.
The sequence is following:

- RolesConfigured
- prepare_2pc() passes
- failover tiriggers
- state changes to ConfiguringRoles
- myrole.apply_config() yields
- commit_2pc() fails

This patch introduces internal method `confapplier.wish_state()` which
tries to satisfy desires state even if it differs at the moment.

I didn't forget about

- [x] Tests
- [x] Changelog
- [x] Documentation

Close #417

### CODE REVIEW

File: cartridge/confapplier.lua [hunk 1]
```diff
@@ -120,8 +125,44 @@ local function set_state(new_state, err)
...
+end
+
+--- Make a wish for meeting desired state.
+-- @function wish_state
+-- @local
+-- @tparam string state
+--   Desired state.
+-- @tretrun string
+--   Final state, may differ from desired.
+local function wish_state(state)
```
Comment (filonenko-mikhail) [inline cartridge/confapplier.lua]: dont you want parametrized deadline ?

File: cartridge/confapplier.lua [hunk 2]
```diff
@@ -120,8 +125,44 @@ local function set_state(new_state, err)
...
+    checks('string')
+
+    if vars.state_notification == nil then
+        vars.state_notification = fiber.cond()
```
Comment (filonenko-mikhail) [inline cartridge/confapplier.lua]: dont you want toplevel init?

File: cartridge/confapplier.lua (hunk 1 above)
Comment (rosik) [reply cartridge/confapplier.lua]: Ok.

File: cartridge/confapplier.lua (hunk 2 above)
Comment (rosik) [reply cartridge/confapplier.lua]: Ok.

**Comment by rosik** [comment]
There is another similar race, which isn't fixed by this patch. Closing in favor of #425

**Pull request closed without merging**